src/compiler/factory.ts

```diff
@@ -66,6 +71,16 @@ namespace ts {
         return clone;
     }
 
+    /**
+     * Creates a deep, memberwise clone of a node with no source map location.
+     */
+    /* @internal */
+    export function getSynthesizedDeepClone<T extends Node>(node: T | undefined): T | undefined {
+        return node

```

Comment:
This whole operation seems somewhat expensive. The call to `visitEachChild` will likely resolve in a call to one of the `updateX` functions in _factory.ts_, which in turn calls `updateNode`, which in turn sets the `original` pointer on the new node, clones the original node's `emitNode`, and sets the `pos` and `end` of the new node. Then `getSynthesizedClone` discards the new object for a fresh one.

If we only need this for entity-name-like expressions, could we instead write a more specific version that doesn't do all of this extra work that is just thrown out? See `serializeQualifiedNameAsExpression` in _ts.ts_ for a function that already does this (albeit with some extra options you don't need).

Reply:
I agree that's it's overkill for this application.  It was actually added for https://github.com/Microsoft/TypeScript/pull/18997 and this was a proof-of-concept application.  If there's a way to make this less expensive, I'd be happy to rework it.

Reply:
Also, consider moving this to _services_ so that it isn't used unintentionally in _compiler_.